[1.95.0] Document provenance and its use in const eval (#707)
ferrocene/specification: src/types-and-traits.rst

Indirection Types
-----------------

.. rubric:: Legality Rules

:dp:`fls_3qI8FXMsyk0f`

> A :t:`pointer type` is a :t:`type` whose :t:`[value]s` indicate memory locations.

Reviewer: Reading through here I think I spotted one scope problem and one naming collision in the new `pointer type` definition, both fixable with the same edit:

1. **Scope.** "A type whose values indicate memory locations" describes function pointer types too — a function pointer's value indicates where code lives, which is exactly why `indirection type` ("a type whose values refer to memory locations") includes them. So as defined, function pointers are `pointer`s, and the provenance rules, the well-formed pointer requirement on const/static values, and the access/offset UB rules all textually apply to them. But whether function pointers have provenance at all is an open question upstream — [rust-lang/unsafe-code-guidelines#340](https://github.com/rust-lang/unsafe-code-guidelines/issues/340), the issue bjorn3 pointed to on Zulip ("References have provenance. I believe it hasn't been decided yet if function pointers have provenance"). The FLS shouldn't answer that question as a side effect of a definition's wording.

2. **Duplication.** `pointer type` and the pre-existing `indirection type` now coexist with definitions that don't distinguish them ("indicate memory locations" vs "refer to memory locations"). Two terms, one meaning, in the same section.

Defining `pointer type` by enumeration fixes both at once, and scopes the term to exactly the two cases that have been decided — references have provenance, and raw pointers uncontroversially do:

```suggestion
A :t:`pointer type` is either a :t:`raw pointer type` or a :t:`reference
type`.
```

(same wording with `:dt:` for the glossary entry).

That keeps the provenance rules from taking a position UCG#340 hasn't, and gives `pointer type` a
meaning `indirection type` doesn't already have. It does leave function-pointer values in const/static final values unconstrained by the FLS — deliberate underspecification until UCG#340 resolves, and worth one line on the record (maybe as an issue we open?) so the gap reads as chosen rather than missed.

Author: Incorporated.
